### ISSUE TITLE
Inconsistent checks for existence of envelope in BaseBinaryStar::ResolveCommonEnvelopeEvent() and *::ResolveEnvelopeLoss()

### ISSUE THREAD

**Issue by jeffriley** [TeamCOMPAS/COMPAS#874]
Inconsistent checks for existence of envelope in BaseBinaryStar::ResolveCommonEnvelopeEvent() and *::ResolveEnvelopeLoss().

BaseBinaryStar::ResolveCommonEnvelopeEvent() determines the values of the 'envelopeFlag1' and 'envelopeFlag2' flags based on the core mass, then calls ResolveEnvelopeLossAndSwitch(), which calls ResolveEnvelopeLoss(TRUE).  Calling ResolveEnvelopeLoss() with the p_NoCheck parameter TRUE bypasses the envelope existence check in ResolveEnvelopeLoss() (though also see issue #873).

So BaseBinaryStar::ResolveCommonEnvelopeEvent() is saying "the star has lost its envelope - update its attributes". But ResolveEnvelopeLoss() will only update the star if it thinks the envelope has been lost, and it uses a different criterion (or criteria if we look at *::ResolveEnvelopeLoss()).

For now it seems not to matter - that we know of. In the test I did it didn't matter (see issue #873), but maybe one day it will... The problem is that we use different criteria to determine if the envelope has been lost. I think if we want to do it properly we would have a function in each stellar class that determines if the envelope has been lost (using the criteria in *::ResolveEnvelopeLoss()) and call that from BaseBinaryStar::ResolveCommonEnvelopeEvent().

I think this came about because some time in the recent past @ilyamandel changed the *::ResolveEnvelopeLoss() functions to use the "correct" core mass (but I might be misremembering that), so the checks there got out of sync with the checks in BaseBinaryStar::ResolveCommonEnvelopeEvent().  Also tagging @avigna because I have a vague recollection (which might also be wrong) the he, at some stage, had something to do with the calculation of the 'envelopeFlag1' and 'envelopeFlag2' flags in BaseBinaryStar::ResolveCommonEnvelopeEvent().

(@ilyamandel and @avigna feel free to unassign yourselves if I'm remembering incorrectly - though your inputs/thoughts would be valuable and welcome).  I'm happy to make the code changes once we decide what's best.

This code is current in COMPAS v02.34.02

**Comment by ilyamandel** [comment]
I am not sure I understand what the issue is.  

ResolveEnvelopeLoss() is intentionally meant to be more generic than just ResolveCommonEnvelopeEvent() -- maybe there is another reason than a CE for why the star lost its envelope and that needs to be resolved.  So I am not too worried about different conditions. 

The  p_NoCheck is something that you like and have introduced in many places, @jeffriley . :-)  But I think what we are doing here is basically enforcing the envelope loss by setting this to true, whether or not there is actually a positive difference between the total mass and the core mass.  So, again, it seems sensible.

In other words, as far as I can tell, both functions correctly implement the behaviour I would have expected, and I don't see a conflict.  What am I missing?

**Comment by jeffriley** [comment]
> The p_NoCheck is something that you like...

Not so @ilyamandel . I hated it when I put it in and I've hated it ever since.  The reason it is there is that in the legacy code there were two snippets of code that did the same thing (basically, what's in *::ResolveEnvelopeLoss(), but one was covered by a check for mass and one was not (I think they were two separate snippets of code in the huge case statement in star.cpp, but they might have been functions - it was a long time ago...). Multiple snippets of code doing pretty-much the same thing invite problems (when someone changes one but forget to change the other), so when I refactored I made it a single function with the p_No_Check parameter.  I didn't like it, but it was the better option.

> ,,,and have introduced in many places

Also not so @ilyamandel.  Unless I'm missing the other places only *::ResolveEnvelopeLoss() has the p_NoCheck parameter (yes, there are several of the functions because there are several stellar types, but it is only that one function (that I know of).

Anyway, I thought the different conditions in ResolveCommonEnvelopeEvent() and ResolveEnvelopeLoss() could cause problems, especially if we remove the p_NoCheck parameter (which is the subject of some testing I did here:  https://github.com/TeamCOMPAS/COMPAS/pull/822#issuecomment-1333173239 - I don't think it is actually required (when I implemented it I meant to check whether it was actually ever required, but never got to it - I did a few weeks ago for that other PR).

But, if you are confident that the checks in ResolveCommonEnvelopeEvent() will *always* be more stringent than the checks in ResolveEnvelopeLoss(), then I think we can close this and remove the p_NoCheck parameter from ResolveEnvelopeLoss() - I don't think it is actually required.

**Comment by ilyamandel** [comment]
Sorry, I was wrong to say that you had introduced p_NoCheck in many places or that you liked it -- I take that back (and apologise)!  I was referring to the general idea of passing a boolean arguments to allow a function to slightly vary its behaviour while avoiding duplication of code.  And I agree it's a good thing!

In any case, while I agree that p_NoCheck is not needed for ResolveCommonEnvelopeEvent() , I am not convinced that p_NoCheck is unnecessary for other calls as described in #873 ; see my comments there.
